db_stress_tool/expected_state.h

```diff
@@ -19,9 +19,181 @@
 #include "rocksdb/file_system.h"
 #include "rocksdb/rocksdb_namespace.h"
 #include "rocksdb/types.h"
+#include "util/autovector.h"
 #include "util/string_util.h"
 
 namespace ROCKSDB_NAMESPACE {
+class ExpectedValue {
+ public:
+  explicit ExpectedValue(std::atomic<uint32_t>* value) : value_(value){};
+  // Requires external locking to prevent concurrent
+  // delete to the same ExpectedValue.
+  bool Exist() const;
+  // Requires external locking to prevent concurrent
+  // write/delete to the same ExpectedValue.
+  void Reset();
+  // Does not require external locking to prevent concurrent
+  // write/delete to the same ExpectedValue.
+  uint32_t Read() const { return value_->load(); }
+  // Requires external locking to prevent concurrent
+  // write/delete to the same ExpectedValue.
+  void Put(bool pending);
+  // Requires external locking to prevent concurrent
+  // write/delete to the same ExpectedValue.
+  uint32_t GetFinalValueBase() const;
+  // Requires external locking to prevent concurrent
+  // write/delete to the same ExpectedValue.
+  bool Delete(bool pending);
+  // Requires external locking to prevent concurrent
+  // write/delete to the same ExpectedValue.
+  void SyncPut(uint32_t value_base);
+  // Requires external locking to prevent concurrent
+  // write/delete to the same ExpectedValue.
+  void SyncSingleDelete();
+
+ private:
+  std::atomic<uint32_t>* value_;
+};
+
+class ExpectedValueHelper {
+ public:
+  // Return whether value is expected not to exist from begining till the end
+  // of the read based on `pre_read_expected_value` and
+  // `pre_read_expected_value`.
+  static bool MustHaveNotExisted(uint32_t pre_read_expected_value,
+                                 uint32_t post_read_expected_value);
+
+  // Return whether value is expected to exist from begining till the end of
+  // the read based on `pre_read_expected_value` and
+  // `pre_read_expected_value`.
+  static bool MustHaveExisted(uint32_t pre_read_expected_value,
+                              uint32_t post_read_expected_value);
+
+  // Return whether the `value_base` falls within the expected value base
+  static bool InExpectedValueBaseRange(uint32_t value_base,
+                                       uint32_t pre_read_expected_value_base,
+                                       uint32_t post_read_expected_value_base,
+                                       bool post_read_expected_pending_write);
```

Comment:
I would use the following function signatures:

```
  static bool MustHaveNotExisted(ExpectedValue pre_read_expected_value,
                                 ExpectedValue post_read_expected_value);

  static bool MustHaveExisted(ExpectedValue pre_read_expected_value,
                              ExpectedValue post_read_expected_value);

  static bool InExpectedValueBaseRange(uint32_t value_base,
                                       ExpectedValue pre_read_expected_value,
                                       ExpectedValue post_read_expected_value);
```

Reply:
Fixed